### DISCUSSION TITLE
Making web adb work with computer's adb.

### DISCUSSION THREAD

**Discussion by pojungc** [yume-chan/ya-webadb#689] [General]
Hi, 
As i understand Tango adb and Google adb don't work well together and if I wanted to use the web adb, I would need to kill the adb server on my computer. 

I wanted to discuss if adding Google adb would ever be possible to Tango? Have you tried integrating it and if you ran into some technical limitations? I would like it be seamless between the web and my computer environment. 

If not possible, would you have have any ideas for workarounds?

**Comment by yume-chan** [comment]
We already support connecting through Google ADB Server: https://docs.tangoapp.dev/tango/server/

The only problem is it's not supported in Web browsers. (A WebSockify-like native app is required to workaround this)

**Reply by pojungc** [reply]
Thanks for the quick response! 
If I understand correctly this native app would sit between adb server and web browser to translate websockets/tcp, is that correct?

Do you know if there are any code reference I could look at to get a better understanding?

**Reply by yume-chan** [reply]
> native app would sit between adb server and web browser to translate websockets/tcp

Correct.

> Do you know if there are any code reference I could look at to get a better understanding?

Our Tango Desktop Bridge is open-source: https://github.com/tango-adb/bridge-rs/blob/main/src/main.rs, it also embeds and starts Google ADB on incoming WebSocket connections. It's not well-documented, you can also refer to projects like https://github.com/novnc/websockify-js as the basic principle is same.

On the Web side, you need to write a custom AdbServerConnector to utilize the WebSocket connection. We don't have a code example for that, but it's very similar to the Node.js TCP socket connector, just converting events to Web Streams: https://github.com/yume-chan/ya-webadb/blob/6a18fa293b3c7d8c89558af39dd332824c7af12e/libraries/adb-server-node-tcp/src/index.ts#L12-L57

If you don't care about compatibility, you can also use the new WebSocketStream API: https://developer.mozilla.org/en-US/docs/Web/API/WebSocketStream. It provides WebSocket connections using Web Streams API natively.

**Reply by pojungc** [reply]
Appreciate the detailed explanation! Will look into this workaround more! 
Btw, love the project and the work that you have done, ty.
